unixfs/tar/writer.go

```diff
@@ -61,91 +64,127 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 type Writer struct {
 	Dag  mdag.DAGService
 	TarW *tar.Writer
+
+	w   io.Writer
+	ctx cxt.Context
+
+	Close      func() error
+	HandleFile func(*mdag.Node, *upb.Data, string) error
+	HandleDir  func(string) error
 }
 
 // NewWriter wraps given io.Writer.
 // compression determines whether to use gzip compression.
-func NewWriter(w io.Writer, dag mdag.DAGService, compression int) (*Writer, error) {
-
+func NewWriter(ctx cxt.Context, dag mdag.DAGService, archive bool, compression int, iw io.Writer) (*Writer, error) {
 	if compression != gzip.NoCompression {
 		var err error
-		w, err = gzip.NewWriterLevel(w, compression)
+		iw, err = gzip.NewWriterLevel(iw, compression)
 		if err != nil {
 			return nil, err
 		}
 	}
-
-	return &Writer{
-		Dag:  dag,
-		TarW: tar.NewWriter(w),
-	}, nil
-}
-
-func (w *Writer) WriteDir(ctx cxt.Context, nd *mdag.Node, fpath string) error {
-	if err := writeDirHeader(w.TarW, fpath); err != nil {
-		return err
+	w := &Writer{
+		Dag: dag,
+		w:   iw,
+		ctx: ctx,
 	}
 
-	for i, ng := range w.Dag.GetDAG(ctx, nd) {
-		child, err := ng.Get(ctx)
-		if err != nil {
+	if !archive && compression != gzip.NoCompression {
+		// gz only case
+		w.HandleFile = func(nd *mdag.Node, pb *upb.Data, fpath string) error {
+			dagr := uio.NewDataFileReader(w.ctx, nd, pb, w.Dag)
+			_, err := dagr.WriteTo(w.w)
 			return err
 		}
 
-		npath := path.Join(fpath, nd.Links[i].Name)
-		if err := w.WriteNode(ctx, child, npath); err != nil {
+		w.HandleDir = func(fpath string) error {
+			return uio.ErrIsDir
+		}
```

Comment:
if this is not taring output, then it shouldn't be part of the `tar.Writer`. 

and this callback stuff is messy. we should be _composing_ writers, instead of putting both functionalities into one function/struct. why isnt this just two different types? composing writers is cheap, and clear.

```go
// a different func in some other package
func getDagWriter(archive bool, compression int, w io.Writer) io.Writer {
  if compression != gzip.NoCompression {
    w = gzip.Writer(w, compression) // only does the gzip stuff
  }

  if archive {
    return tar.Writer(w, compression) // only does the tar stuff
  }

  return  // what goes here? not archiving nor compressing. 
}
```